File: clients/py/kascov.py
"""kascov.py — a tiny zero-dependency client for the kascov JSON API.

Python 3.9+, stdlib only (urllib). CORS-open API, no keys.

    from kascov import Kascov
    k = Kascov("testnet-10")
    page = k.coins(limit=100)
    coin = k.coin(page["covenants"][0]["covenant_id"])
    for ev in k.stream():           # live events (SSE), blocks forever
        print(ev["kind"], ev["covenant_id"])
    for ev in k.stream(covenant=cid):   # one coin only

The API needs no token. An optional lane token (minted at kascov.io/lane)
rides along as an X-Kascov-Lane header on every request; it buys extra
capacity on the holder lane and nothing else — the anonymous tier keeps
working without it:

    k = Kascov("mainnet", lane_token="...")

Passport badges verify locally — no request, no trust in the server:

    from kascov import verify_badge
    verify_badge(claim, proof, root)  # True only if the claim is in the tree

Publishing to PyPI is a separate decision — this file is the whole client.
"""
from __future__ import annotations

import hashlib
import json
import re
import urllib.parse
import urllib.request
from typing import Any, Dict, Iterator, Optional, Sequence

DEFAULT_BASE = "https://kascov.io"


class Kascov:
    def __init__(
        self,
        network: str = "mainnet",
        base: str = DEFAULT_BASE,
        lane_token: Optional[str] = None,
    ) -> None:
        self.network = network
        self.base = base.rstrip("/")
        self.lane_token = lane_token

    def _headers(self, accept: str) -> Dict[str, str]:
        # every request goes through here so the lane header can never be
        # forgotten on one endpoint and sent on another
        h = {"accept": accept, "user-agent": "kascov-py"}
        if self.lane_token:
            h["X-Kascov-Lane"] = self.lane_token
        return h

    def _get(self, path: str) -> Dict[str, Any]:
        req = urllib.request.Request(
            f"{self.base}{path}", headers=self._headers("application/json")
        )
        with urllib.request.urlopen(req, timeout=60) as res:
            return json.load(res)

    def _get_with_query(self, path: str, params: Dict[str, Any]) -> Dict[str, Any]:
        clean = {}
        for key, value in params.items():
            if value is not None:
                clean[key] = str(value).lower() if isinstance(value, bool) else value
        qs = f"?{urllib.parse.urlencode(clean)}" if clean else ""
        return self._get(f"{path}{qs}")

    def live(self) -> Dict[str, Any]:
        """Small fast feed: stats + chain tip + newest ~150 events."""
        return self._get(f"/data/{self.network}-live.json")

    def coins(
        self,
        limit: Optional[int] = None,
        after_daa: Optional[int] = None,
        after_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """One page of coin summaries, newest first. Pass the previous page's
        next_after_daa / next_after_id to walk older coins."""
        q = {k: v for k, v in {"limit": limit, "after_daa": after_daa, "after_id": after_id}.items() if v is not None}
        qs = f"?{urllib.parse.urlencode(q)}" if q else ""
        return self._get(f"/data/{self.network}.json{qs}")

    def coin(self, covenant_id: str) -> Dict[str, Any]:
        """One coin's full story: events, UTXOs (scripts/reveals), holders."""
        return self._get(f"/data/{self.network}/c/{covenant_id}.json")

    def tx(self, txid: str) -> Dict[str, Any]:
        """Which covenant(s) did this transaction move?"""
        return self._get(f"/data/{self.network}/tx/{txid}.json")

    def address(self, addr_or_pubkey: str) -> Dict[str, Any]:
        """Smart coins an address/pubkey funded, received, or controls."""
        return self._get(f"/data/{self.network}/addr/{urllib.parse.quote(addr_or_pubkey)}.json")

    def digest(self) -> Dict[str, Any]:
        """Last-24h digest: births/moves/burns, value born, headliners."""
        return self._get(f"/data/{self.network}/digest.json")

    def galaxy(self) -> Dict[str, Any]:
        """The whole-network app graph (positions + weighted edges)."""
        return self._get(f"/data/{self.network}/galaxy.json")

    def reorgs(self) -> Dict[str, Any]:
        """Recent chain reorgs the indexer rolled back through."""
        return self._get(f"/data/{self.network}/reorgs.json")

    def stream_info(self) -> Dict[str, Any]:
        """Durable delivery bounds for snapshot-to-stream handoff."""
        return self._get(f"/data/{self.network}/stream-info.json")

    def events(
        self,
        after: Optional[str] = None,
        limit: Optional[int] = None,
        covenant: Optional[str] = None,
        application: Optional[str] = None,
        artifact: Optional[str] = None,
        actor: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Read one durable delivery page in global cursor order."""
        q = {
            key: value
            for key, value in {
                "after": after,
                "limit": limit,
                "covenant": covenant,
                "application": application,
                "artifact": artifact,
                "actor": actor,
            }.items()
            if value is not None
        }
        suffix = f"?{urllib.parse.urlencode(q)}" if q else ""
        return self._get(f"/data/{self.network}/events{suffix}")

    def application_state(self, application: str, **filters: Any) -> Dict[str, Any]:
        """Current accepted application state with cursor and freshness metadata."""
        q = {key: value for key, value in filters.items() if value is not None}
        suffix = f"?{urllib.parse.urlencode(q)}" if q else ""
        application = urllib.parse.quote(application, safe="")
        return self._get(f"/data/{self.network}/apps/{application}/state{suffix}")

    def templates(self) -> Dict[str, Any]:
        """Contract-type analytics (what's running on this network)."""
        return self._get(f"/data/{self.network}/templates.json")

    def tokens(
        self, limit=None, after_daa=None, after_id=None, status=None,
        phase=None, kind=None, q=None,
    ) -> Dict[str, Any]:
        """Derived token/minter directory. Any option opts into a bounded page."""
        return self._get_with_query(f"/data/{self.network}/tokens.json", {
            "limit": limit, "after_daa": after_daa, "after_id": after_id,
            "status": status, "phase": phase, "kind": kind, "q": q,
        })

    def token(
        self, covenant_id: str, limit=None, events_limit=None,
        after_seq=None, before_seq=None, order=None,
    ) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/token/{covenant_id}", {
            "limit": limit, "events_limit": events_limit, "after_seq": after_seq,
            "before_seq": before_seq, "order": order,
        })

    def token_holders(self, covenant_id: str, limit=None, after_balance=None, after_owner=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/token/{covenant_id}/holders", {
            "limit": limit, "after_balance": after_balance, "after_owner": after_owner,
        })

    def token_events(self, covenant_id: str, limit=None, after_seq=None, before_seq=None, order=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/token/{covenant_id}/events", {
            "limit": limit, "after_seq": after_seq, "before_seq": before_seq, "order": order,
        })

    def token_trades(self, covenant_id: str, limit=None, before_seq=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/token/{covenant_id}/trades", {
            "limit": limit, "before_seq": before_seq,
        })

    def trades(
        self, limit=None, token_id=None, market_id=None, side=None,
        before_daa=None, before_token=None, before_seq=None,
    ) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/trades", {
            "limit": limit, "token_id": token_id, "market_id": market_id,
            "side": side, "before_daa": before_daa,
            "before_token": before_token, "before_seq": before_seq,
        })

    def markets(self, limit=None, after_id=None, phase=None, priced=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/markets", {
            "limit": limit, "after_id": after_id, "phase": phase, "priced": priced,
        })

    def market(self, covenant_id: str) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/market/{covenant_id}")

    def token_market(self, covenant_id: str) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/token/{covenant_id}/market")

    def pools(self, limit=None, after_id=None, priced=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/pools", {
            "limit": limit, "after_id": after_id, "priced": priced,
        })

    def pool(self, covenant_id: str) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/pool/{covenant_id}")

    def vesting(self, limit=None, after_id=None) -> Dict[str, Any]:
        return self._get_with_query(f"/data/{self.network}/vesting", {
            "limit": limit, "after_id": after_id,
        })

    def vesting_detail(self, token_or_lock_id: str) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/vesting/{token_or_lock_id}")

    def vesting_claims(self, token_or_lock_id: str) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/vesting/{token_or_lock_id}/claims")

    def index(self) -> Dict[str, Any]:
        return self._get(f"/data/{self.network}/index.json")

    def openapi(self) -> Dict[str, Any]:
        return self._get("/openapi.json")

    def activity(self, range: str = "24h") -> Dict[str, Any]:
        """Births/moves/burns per DAA bucket. range: 1h|6h|24h|48h|all"""
        return self._get(f"/data/{self.network}/activity.json?range={range}")

    def stream(
        self,
        after: Optional[str] = None,
        covenant: Optional[str] = None,
        application: Optional[str] = None,
        artifact: Optional[str] = None,
        actor: Optional[str] = None,
        reconnect: bool = True,
    ) -> Iterator[Dict[str, Any]]:
        """Durable SSE events. Reconnects with Last-Event-ID after the first response.

        Each result keeps the server payload and adds ``_event`` and ``_cursor``.
        A ``reset`` result ends the iterator so the caller can reload a snapshot.
        """
        q = {
            key: value
            for key, value in {
                "after": after,
                "covenant": covenant,
                "application": application,
                "artifact": artifact,
                "actor": actor,
            }.items()
            if value is not None
        }
        suffix = f"?{urllib.parse.urlencode(q)}" if q else ""
        url = f"{self.base}/data/{self.network}/stream{suffix}"
        cursor: Optional[str] = None
        while True:
            headers = {"accept": "text/event-stream", "user-agent": "kascov-py"}
            if cursor is not None:
                headers["Last-Event-ID"] = cursor
            req = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(req, timeout=None) as res:
                event_name = "message"
                event_id: Optional[str] = None
                data: list[str] = []
                for raw in res:
                    line = raw.decode("utf-8", "replace").rstrip("\r\n")
                    if line == "":
                        if data:
                            try:
                                payload = json.loads("\n".join(data))
                            except json.JSONDecodeError:
                                payload = None
                            if isinstance(payload, dict):
                                if event_id is not None:
                                    cursor = event_id
                                payload["_event"] = event_name
                                payload["_cursor"] = event_id
                                yield payload
                                if event_name == "reset":
                                    return
                        event_name, event_id, data = "message", None, []
                    elif line.startswith("event:"):
                        event_name = line[6:].strip()
                    elif line.startswith("id:"):
                        event_id = line[3:].strip()
                    elif line.startswith("data:"):
                        data.append(line[5:].lstrip())
            if not reconnect:
                return


# ---------------------------------------------------------------------------
# Passport badge verification — pure and local. The scheme (which the bot's
# merkle publisher in scripts/discord-holder-bot.mjs must match exactly — the
# publisher was not yet written when this landed, so this file is the spec
# both sides pin):
#
#     leaf  = sha256(canonical JSON of the claim)   # keys sorted, no spaces
#     node  = sha256(lo + hi)                       # the PAIR sorted as bytes
#     root  = the published 32-byte hex string
#
# Claims stick to strings, integers, booleans and None — floats serialize
# differently across languages and would fork the leaf hash.

_HEX64 = re.compile(r"^[0-9a-f]{64}$")


def canonical_json(value: Any) -> str:
    """Canonical JSON: recursively sorted keys, no whitespace, raw unicode.
    Matches the js client's canonicalJson for claims built from the types
    listed above."""
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def verify_badge(claim: Any, proof: Sequence[str], root: str) -> bool:
    """Verify a passport badge against a published merkle root, entirely
    locally. claim: the claim object as published; proof: sibling hashes
    leaf->root (64-char hex each); root: the published root (64-char hex).
    An empty proof means the claim IS the whole tree. Malformed input returns
    False — the verifier fails closed, it never raises."""
    if not isinstance(root, str) or not isinstance(proof, (list, tuple)):
        return False
    want = root.lower()
    if not _HEX64.fullmatch(want):
        return False
    cur = hashlib.sha256(canonical_json(claim).encode("utf-8")).hexdigest()
    for step in proof:
        if not isinstance(step, str):
            return False
        sib = step.lower()
        if not _HEX64.fullmatch(sib):
            return False
        # pair-sorted concatenation: lexicographic hex order == byte order here
        lo, hi = sorted((cur, sib))
        cur = hashlib.sha256(bytes.fromhex(lo) + bytes.fromhex(hi)).hexdigest()
    return cur == want

File: clients/js/kascov.mjs
/* kascov.mjs — a tiny zero-dependency client for the kascov JSON API.
   Works in Node 18+ and the browser (native fetch). CORS is open, no keys.

     import { Kascov } from './kascov.mjs';
     const k = new Kascov('testnet-10');
     const { covenants, next_after_daa } = await k.coins({ limit: 100 });
     const coin = await k.coin(covenants[0].covenant_id);
     const stream = k.stream({ onMessage: (ev) => console.log(ev.kind) });


   Publishing to npm is a separate decision — this file is the whole client. */

const DEFAULT_BASE = 'https://kascov.io';

export class Kascov {
  constructor(network = 'mainnet', base = DEFAULT_BASE, { laneToken = '' } = {}) {
    this.network = network;
    this.base = base.replace(/\/$/, '');
    this.laneToken = laneToken;
  }

  /* every request goes through here so the lane header can never be forgotten
     on one endpoint and sent on another */
  #headers(accept) {
    const h = { accept };
    if (this.laneToken) h['x-kascov-lane'] = this.laneToken;
    return h;
  }

  async #get(path) {
    const res = await fetch(`${this.base}${path}`, { headers: this.#headers('application/json') });
    if (!res.ok) throw new Error(`kascov: ${path} → HTTP ${res.status}`);
    return res.json();
  }

  #getWithQuery(path, entries) {
    const q = new URLSearchParams();
    for (const [key, value] of entries) {
      if (value != null) q.set(key, value);
    }
    const suffix = q.toString();
    return this.#get(`${path}${suffix ? `?${suffix}` : ''}`);
  }

  /** Small fast feed: stats + chain tip + newest ~150 events. Poll this. */
  live() { return this.#get(`/data/${this.network}-live.json`); }

  /** One page of coin summaries, newest activity first.
      opts: { limit, afterDaa, afterId } — pass the previous page's
      next_after_daa / next_after_id to walk older coins. */
  coins(opts = {}) {
    const q = new URLSearchParams();
    if (opts.limit != null) q.set('limit', opts.limit);
    if (opts.afterDaa != null) q.set('after_daa', opts.afterDaa);
    if (opts.afterId != null) q.set('after_id', opts.afterId);
    const s = q.toString();
    return this.#get(`/data/${this.network}.json${s ? `?${s}` : ''}`);
  }

  /** One coin's full story: events (payloads, moved-with), UTXOs (scripts,
      reveals, budgets), holders. */
  coin(covenantId) { return this.#get(`/data/${this.network}/c/${covenantId}.json`); }

  /** Which covenant(s) did this transaction move? */
  tx(txid) { return this.#get(`/data/${this.network}/tx/${txid}.json`); }

  /** Smart coins an address/pubkey funded, received, or controls. */
  address(addrOrPubkey) { return this.#get(`/data/${this.network}/addr/${encodeURIComponent(addrOrPubkey)}.json`); }

  /** Last-24h digest: births/moves/burns, value born, headliner coins. */
  digest() { return this.#get(`/data/${this.network}/digest.json`); }

  /** The whole-network app graph (positions + weighted edges). */
  galaxy() { return this.#get(`/data/${this.network}/galaxy.json`); }

  /** Recent chain reorgs the indexer rolled back through. */
  reorgs() { return this.#get(`/data/${this.network}/reorgs.json`); }

  /** Durable delivery bounds for snapshot-to-stream handoff. */
  streamInfo() { return this.#get(`/data/${this.network}/stream-info.json`); }

  /** One durable delivery page in global cursor order.
      opts: { after, limit, covenant, application, artifact, actor }. */
  events(opts = {}) {
    const q = new URLSearchParams();
    for (const [key, value] of Object.entries(opts)) {
      if (value != null) q.set(key, value);
    }
    const suffix = q.toString();
    return this.#get(`/data/${this.network}/events${suffix ? `?${suffix}` : ''}`);
  }

  /** Current accepted application state, with cursor and freshness metadata. */
  applicationState(application, opts = {}) {
    const q = new URLSearchParams();
    for (const [key, value] of Object.entries(opts)) {
      if (value != null) q.set(key, value);
    }
    const suffix = q.toString();
    return this.#get(`/data/${this.network}/apps/${encodeURIComponent(application)}/state${suffix ? `?${suffix}` : ''}`);
  }

  /** Contract-type analytics (what's running on this network). */
  templates() { return this.#get(`/data/${this.network}/templates.json`); }

  /** Derived token/minter directory. No options preserves the legacy full list;
      any option opts into a bounded page. */
  tokens(opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/tokens.json`, [
      ['limit', opts.limit], ['after_daa', opts.afterDaa], ['after_id', opts.afterId],
      ['status', opts.status], ['phase', opts.phase], ['kind', opts.kind], ['q', opts.q],
    ]);
  }

  /** One derived token, with bounded embedded holders and events. */
  token(covenantId, opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/token/${covenantId}`, [
      ['limit', opts.limit], ['events_limit', opts.eventsLimit],
      ['after_seq', opts.afterSeq], ['before_seq', opts.beforeSeq], ['order', opts.order],
    ]);
  }

  tokenHolders(covenantId, opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/token/${covenantId}/holders`, [
      ['limit', opts.limit], ['after_balance', opts.afterBalance], ['after_owner', opts.afterOwner],
    ]);
  }

  tokenEvents(covenantId, opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/token/${covenantId}/events`, [
      ['limit', opts.limit], ['after_seq', opts.afterSeq],
      ['before_seq', opts.beforeSeq], ['order', opts.order],
    ]);
  }

  tokenTrades(covenantId, opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/token/${covenantId}/trades`, [
      ['limit', opts.limit], ['before_seq', opts.beforeSeq],
    ]);
  }

  trades(opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/trades`, [
      ['limit', opts.limit], ['token_id', opts.tokenId], ['market_id', opts.marketId],
      ['side', opts.side], ['before_daa', opts.beforeDaa],
      ['before_token', opts.beforeToken], ['before_seq', opts.beforeSeq],
    ]);
  }

  markets(opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/markets`, [
      ['limit', opts.limit], ['after_id', opts.afterId],
      ['phase', opts.phase], ['priced', opts.priced],
    ]);
  }

  market(covenantId) { return this.#get(`/data/${this.network}/market/${covenantId}`); }
  tokenMarket(covenantId) { return this.#get(`/data/${this.network}/token/${covenantId}/market`); }

  pools(opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/pools`, [
      ['limit', opts.limit], ['after_id', opts.afterId], ['priced', opts.priced],
    ]);
  }

  pool(covenantId) { return this.#get(`/data/${this.network}/pool/${covenantId}`); }
  vesting(opts = {}) {
    return this.#getWithQuery(`/data/${this.network}/vesting`, [
      ['limit', opts.limit], ['after_id', opts.afterId],
    ]);
  }
  vestingDetail(tokenOrLockId) { return this.#get(`/data/${this.network}/vesting/${tokenOrLockId}`); }
  vestingClaims(tokenOrLockId) { return this.#get(`/data/${this.network}/vesting/${tokenOrLockId}/claims`); }
  index() { return this.#get(`/data/${this.network}/index.json`); }
  openapi() { return this.#get('/openapi.json'); }

  /** Births/moves/burns per DAA bucket. range: 1h|6h|24h|48h|all */
  activity(range = '24h') { return this.#get(`/data/${this.network}/activity.json?range=${range}`); }

  /** Open a native EventSource. The browser sends Last-Event-ID on automatic
      reconnect. A reset closes the source so the caller can load a snapshot
      before opening a new stream. */
  stream(opts = {}) {
    const EventSourceClass = opts.EventSource || globalThis.EventSource;
    if (!EventSourceClass) throw new Error('kascov: EventSource is unavailable');
    const q = new URLSearchParams();
    for (const key of ['after', 'covenant', 'application', 'artifact', 'actor']) {
      if (opts[key] != null) q.set(key, opts[key]);

    }
    const suffix = q.toString();
    const source = new EventSourceClass(
      `${this.base}/data/${this.network}/stream${suffix ? `?${suffix}` : ''}`,
    );
    source.onopen = (event) => opts.onOpen?.(event);
    source.onerror = (event) => opts.onError?.(event);
    const deliver = (event) => {
      try { opts.onMessage?.(JSON.parse(event.data), event); } catch { /* invalid data is ignored */ }
    };
    source.onmessage = deliver;
    for (const kind of ['accepted', 'removed', 'projection_repaired', 'checkpoint']) {
      source.addEventListener(kind, deliver);
    }
    source.addEventListener('reset', (event) => {
      let reset = null;
      try { reset = JSON.parse(event.data); } catch { /* malformed reset stays null */ }
      source.close();
      opts.onReset?.(reset, event);
    });
    return source;
  }
}

/* ------------------------------------------------------------------------
   Passport badge verification — pure and local. The scheme (which the bot's
   merkle publisher in scripts/discord-holder-bot.mjs must match exactly —
   the publisher was not yet written when this landed, so this file is the
   spec both sides pin):

     leaf  = sha256(canonical JSON of the claim)   // keys sorted, no spaces
     node  = sha256(lo ++ hi)                      // the PAIR sorted as bytes
     root  = the published 32-byte hex string

   Claims stick to strings, integers, booleans and null — floats serialize
   differently across languages and would fork the leaf hash. */

/** Canonical JSON: recursively sorted keys, no whitespace, raw unicode.
    Matches python's json.dumps(sort_keys=True, separators=(",", ":"),
    ensure_ascii=False) for claims built from the types listed above. */
export function canonicalJson(value) {
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
  return `{${Object.keys(value).sort()
    .map((k) => `${JSON.stringify(k)}:${canonicalJson(value[k])}`).join(',')}}`;
}

// sha256 in plain js so the module keeps its browser promise (a top-level
// node:crypto import would break a native browser import) and verification
// stays synchronous. cross-checked against node:crypto in the tests.
const K = [
  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
  0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
  0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
  0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
  0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
  0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
];
const rotr = (x, n) => ((x >>> n) | (x << (32 - n))) >>> 0;

/** sha256 as lowercase hex. Accepts a string (UTF-8 encoded) or Uint8Array. */
export function sha256Hex(input) {
  const bytes = typeof input === 'string' ? new TextEncoder().encode(input) : input;
  // pad to a 64-byte multiple: 0x80, zeros, 64-bit big-endian bit length
  const bitLen = bytes.length * 8;
  const padded = new Uint8Array((((bytes.length + 8) >> 6) << 6) + 64);
  padded.set(bytes);
  padded[bytes.length] = 0x80;
  const dv = new DataView(padded.buffer);
  dv.setUint32(padded.length - 8, Math.floor(bitLen / 4294967296));
  dv.setUint32(padded.length - 4, bitLen >>> 0);
  const h = [0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19];
  const w = new Uint32Array(64);
  for (let off = 0; off < padded.length; off += 64) {
    for (let i = 0; i < 16; i++) w[i] = dv.getUint32(off + i * 4);
    for (let i = 16; i < 64; i++) {
      const s0 = rotr(w[i - 15], 7) ^ rotr(w[i - 15], 18) ^ (w[i - 15] >>> 3);
      const s1 = rotr(w[i - 2], 17) ^ rotr(w[i - 2], 19) ^ (w[i - 2] >>> 10);
      w[i] = (w[i - 16] + s0 + w[i - 7] + s1) >>> 0;
    }
    let [a, b, c, d, e, f, g, hh] = h;
    for (let i = 0; i < 64; i++) {
      const t1 = (hh + (rotr(e, 6) ^ rotr(e, 11) ^ rotr(e, 25)) + ((e & f) ^ (~e & g)) + K[i] + w[i]) >>> 0;
      const t2 = ((rotr(a, 2) ^ rotr(a, 13) ^ rotr(a, 22)) + ((a & b) ^ (a & c) ^ (b & c))) >>> 0;
      hh = g; g = f; f = e; e = (d + t1) >>> 0; d = c; c = b; b = a; a = (t1 + t2) >>> 0;
    }
    [a, b, c, d, e, f, g, hh].forEach((v, i) => { h[i] = (h[i] + v) >>> 0; });
  }
  return h.map((v) => v.toString(16).padStart(8, '0')).join('');
}

const HEX64 = /^[0-9a-f]{64}$/;
const hexToBytes = (hex) => {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  return out;
};

/** Verify a passport badge against a published merkle root, entirely locally.
    claim: the claim object as published; proof: sibling hashes leaf→root
    (64-char hex each); root: the published root (64-char hex). An empty proof
    means the claim IS the whole tree. Malformed input returns false — the
    verifier fails closed, it never throws. */
export function verifyBadge(claim, proof, root) {
  if (!Array.isArray(proof) || typeof root !== 'string') return false;
  const want = root.toLowerCase();
  if (!HEX64.test(want)) return false;
  let cur = sha256Hex(canonicalJson(claim));
  for (const step of proof) {
    if (typeof step !== 'string') return false;
    const sib = step.toLowerCase();
    if (!HEX64.test(sib)) return false;
    // pair-sorted concatenation: lexicographic hex order == byte order here
    const [lo, hi] = cur <= sib ? [cur, sib] : [sib, cur];
    cur = sha256Hex(hexToBytes(lo + hi));
  }
  return cur === want;
}

export default Kascov;
